File: sdks/kotlin/core/src/main/kotlin/io/getstream/visionagents/core/Backend.kt
package io.getstream.visionagents.core

import io.ktor.client.HttpClient
import io.ktor.client.engine.okhttp.OkHttp
import io.ktor.client.plugins.websocket.WebSockets
import io.ktor.http.URLBuilder
import io.ktor.http.URLProtocol
import io.ktor.http.appendPathSegments
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import okhttp3.OkHttpClient

/**
 * Somebody a device is acting for, as Stream knows them.
 *
 * Only the id reaches the router; the token is what proves it. The name and image are carried
 * so chat and video have something to show without a second lookup.
 */
public data class User(
    val id: String,
    val name: String = "",
    val image: String = "",
)

/**
 * Hands over a token for the user, and a fresh one when asked again.
 *
 * A provider rather than a string, because a token expires and an hour-long conversation
 * should not end when it does. It is asked once, then again after a 401.
 */
public fun interface TokenProvider {
    public suspend fun token(): String
}

/**
 * Where the router is and who is asking.
 *
 * A phone holds no secret worth having, so there is no API secret here. A router with nothing
 * in front of it is reached by [customerId]; a deployment verifying tokens is reached by
 * [apiKey] and the token [setUser] is given, which the app's own backend minted.
 *
 * Constructing one does no I/O. Pass an [OkHttpClient] to share the app's connection pool,
 * proxy and certificate pinning; [close] releases the HTTP client built around it.
 */
public class Backend(
    url: String,
    public val customerId: String = "",
    public val apiKey: String = "",
    okHttpClient: OkHttpClient? = null,
) : AutoCloseable {
    /** The router's base URL, with no trailing slash. */
    public val url: String = url.trimEnd('/')

    internal val http: HttpClient = HttpClient(OkHttp) {
        expectSuccess = false
        install(WebSockets)
        engine {
            if (okHttpClient != null) preconfigured = okHttpClient
        }
    }

    private val lock = Mutex()
    @Volatile private var identity: Identity? = null
    @Volatile private var cached: String? = null

    /** Who this is acting for, or null until [setUser]. */
    public val user: User? get() = identity?.user

    /**
     * Says who this device is acting for, and how to prove it.
     *
     * Against a router reached by customer id the token is not read, and the user id is what
     * names the end user; behind a verifying deployment the token is the whole of it.
     */
    public fun setUser(user: User, token: TokenProvider) {
        if (user.id.isEmpty()) throw AgentsException.Configuration("a user needs an id")
        identity = Identity(user, token)
        cached = null
    }

    /** Says who this device is acting for, with a token that will not be refreshed. */
    public fun setUser(user: User, token: String) {
        setUser(user, TokenProvider { token })
    }

    /** Acts as a guest from here on. */
    public fun setUser(guest: Guest) {
        setUser(User(guest.id, guest.name), guest.token)
    }

    /** Forgets the user, which is what signing out is. */
    public fun clearUser() {
        identity = null
        cached = null
    }

    /**
     * The headers every request and every socket handshake carries.
     *
     * `Stream-Auth-Type: jwt` says this caller is somebody's device rather than their backend.
     * It is what makes the router refuse the paths that configure an agent, and it is sent
     * even to a router with nothing in front of it, which would otherwise take the caller for
     * a backend.
     */
    internal suspend fun headers(): Map<String, String> {
        val headers = linkedMapOf("Stream-Auth-Type" to "jwt")
        val who = identity
        if (apiKey.isNotEmpty()) {
            if (who == null) {
                throw AgentsException.Configuration(
                    "an api key needs a user token to go with it; call setUser first",
                )
            }
            headers["X-Api-Key"] = apiKey
            headers["Authorization"] = "Bearer ${token(who)}"
            return headers
        }
        if (customerId.isEmpty()) {
            throw AgentsException.Configuration("pass a customerId, or an apiKey and setUser")
        }
        headers["X-Customer-Id"] = customerId
        if (who != null) headers["X-Stream-User-Id"] = who.user.id
        return headers
    }

    /**
     * Drops the token held, so the next request asks the provider again. Reports whether
     * there was a provider to ask, which is whether a retry could go any differently.
     */
    internal fun expireToken(): Boolean {
        cached = null
        return apiKey.isNotEmpty() && identity != null
    }

    /** The URL for a path under the router, with its segments encoded. */
    internal fun url(vararg segments: String, query: List<Pair<String, String>> = emptyList()): String =
        URLBuilder(url).apply {
            appendPathSegments(segments.toList(), encodeSlash = true)
            query.forEach { (name, value) -> parameters.append(name, value) }
        }.buildString()

    /**
     * The socket URL for a path under the router.
     *
     * Credentials go in the handshake's headers rather than the query string, which a phone
     * can do and a browser cannot: a token in a URL ends up in every log that URL passes
     * through.
     */
    internal fun socketUrl(vararg segments: String, query: List<Pair<String, String>>): String =
        URLBuilder(url(*segments, query = query)).apply {
            protocol = if (protocol == URLProtocol.HTTPS) URLProtocol.WSS else URLProtocol.WS
        }.buildString()

    override fun close() {
        http.close()
    }

    private suspend fun token(who: Identity): String {
        cached?.let { return it }
        return lock.withLock {
            // Asked once however many requests were waiting for it.
            cached ?: who.token.token().also { if (identity === who) cached = it }
        }
    }

    private class Identity(val user: User, val token: TokenProvider)
}

File: sdks/dart/core/lib/src/backend.dart
/// Hands over the token to present, asked for on every request and socket handshake.
///
/// A function rather than a value so a long-lived client never wakes up holding an expired
/// token. Caching one until it is about to expire is the provider's business.
typedef TokenProvider = Future<String> Function();

/// Where the router is and who is asking.
///
/// A device holds no secret worth having, so there is no API secret here. There are two ways
/// to be let in, and which one a deployment takes is its `ROUTER_AUTH_MODE`, not a choice
/// made per request:
///
/// * `customerId`, for a router in `noauth` mode, which believes the header. `userId` names
///   the end user, since there is no token to name them.
/// * `apiKey` with a `token` a backend minted for this user, or a guest's token, for a router
///   verifying tokens.
///
/// Every request and handshake is sent as a device's: `Stream-Auth-Type: jwt` on requests,
/// and no auth type at all on a socket, which the router reads the same way.
final class Backend {
  Backend({
    required this.url,
    this.customerId = '',
    this.apiKey = '',
    this.userId = '',
    this.token,
  }) {
    if (customerId.isEmpty && apiKey.isEmpty) {
      throw ArgumentError('a backend needs a customerId, or an apiKey with a token');
    }
    if (apiKey.isNotEmpty && token == null) {
      throw ArgumentError('an apiKey is only half a credential: pass the token as well');
    }
  }

  /// The router's base URL.
  final Uri url;

  /// Which tenant's agents these are, for a router that believes the header.
  final String customerId;

  /// The public half of a credential, for a router verifying tokens.
  final String apiKey;

  /// Who is asking, for a router that believes the header. With a token the token says so.
  final String userId;

  final TokenProvider? token;

  /// The same router, asked by somebody else.
  Backend withUser(String userId, {TokenProvider? token}) => Backend(
    url: url,
    customerId: customerId,
    apiKey: apiKey,
    userId: userId,
    token: token ?? this.token,
  );

  /// What every request carries.
  ///
  /// `Stream-Auth-Type: jwt` declares a device, and is sent even to an open router, which
  /// would otherwise take a caller with no proxy in front of it for a backend.
  Future<Map<String, String>> headers() async {
    if (apiKey.isEmpty) {
      return {
        'X-Customer-Id': customerId,
        'Stream-Auth-Type': 'jwt',
        if (userId.isNotEmpty) 'X-Stream-User-Id': userId,
      };
    }
    return {
      'X-Api-Key': apiKey,
      'Authorization': 'Bearer ${await token!()}',
      'Stream-Auth-Type': 'jwt',
    };
  }

  /// The URL for a request path, with the query values that are set.
  Uri requestUri(String path, [Map<String, String?> query = const {}]) {
    final present = {for (final MapEntry(:key, :value) in query.entries) key: ?value};
    return url.replace(
      path: _join(url.path, path),
      queryParameters: present.isEmpty ? null : present,
    );
  }

  /// The URL of a socket under the router, credentials included.
  ///
  /// In the query because a browser WebSocket cannot set headers, and one URL shape for every
  /// platform is worth more than keeping a token out of a URL the VM could have avoided.
  /// There is deliberately no auth type here: the spec gives it no query counterpart, so a
  /// socket cannot claim to be a backend.
  Future<Uri> socketUri(String path, [Map<String, String> query = const {}]) async {
    final credentials = apiKey.isEmpty
        ? {'customer_id': customerId, if (userId.isNotEmpty) 'user_id': userId}
        : {'api_key': apiKey, 'token': await token!()};
    return url.replace(
      scheme: url.scheme == 'https' ? 'wss' : 'ws',
      path: _join(url.path, path),
      queryParameters: {...query, ...credentials},
    );
  }
}

String _join(String base, String path) =>
    base.endsWith('/') ? '${base.substring(0, base.length - 1)}$path' : '$base$path';
